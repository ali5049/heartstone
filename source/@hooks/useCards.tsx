import { useNavigation } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { useCallback, useEffect, useState } from 'react';
import { ICards } from '../@types/card.types';
import { RootStackParamList } from '../@types/navigation.types';
import axiosInstance from '../Services/axiosInterceptor';
import { getCardsByType, getUniqueTypes } from '../utils/helpers';

type CardsScreenNavigationProp = NativeStackNavigationProp<RootStackParamList, 'Cards'>;

const useCards = () => {

    const [cardTypes, setCardTypes] = useState<string[]>([]);
    const [cards, setCards] = useState<ICards>({});
    const [isLoading, setIsLoading] = useState<boolean>(false);
    const navigation = useNavigation<CardsScreenNavigationProp>();

    // Navigate to cards screen on card type press
    const onCardTypePress = (type: string) => {
        //Filter the cards based on selected card type
        const filteredCards = getCardsByType(type, cards);
        navigation.navigate('Cards', { cards: filteredCards });
    };

    // Fetch all the cards from rapid api
    const fetchCards = useCallback(async () => {
        setIsLoading(true);
        try {
            const response = await axiosInstance({
                url: 'cards',
                method: 'GET',
            });
            setIsLoading(false);
            setCards(response?.data);

            //Get the unique card types to display on card types screen
            const _types = getUniqueTypes(response?.data);
            console.log('unique types', _types);
            setCardTypes(_types);
        }
        catch (err) {
            console.log('error is', err);
        }
        setIsLoading(false);
    }, []);


    useEffect(() => {
        fetchCards();
    }, [fetchCards]);

    return {
        cardTypes,
        onCardTypePress,
        isLoading,
    };

};

export { useCards };


