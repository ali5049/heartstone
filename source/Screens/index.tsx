export { default as CardTypes } from './CardTypes';
export { default as Cards } from './Cards';
