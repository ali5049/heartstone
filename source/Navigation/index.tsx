import * as React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {createNativeStackNavigator} from '@react-navigation/native-stack';
import {Cards, CardTypes} from '../Screens';
import {RootStackParamList} from '../@types/navigation.types';

const Stack = createNativeStackNavigator<RootStackParamList>();

const Navigation = (): React.JSX.Element => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen name="CardTypes" component={CardTypes} />
        <Stack.Screen name="Cards" component={Cards} />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default Navigation;
