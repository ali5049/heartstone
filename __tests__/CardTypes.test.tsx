
import React from 'react';
import { render, waitFor } from '@testing-library/react-native';
import {CardTypes} from '../source/Screens';
import * as useCardsModule from '../source/@hooks/useCards';

jest.mock('../source/@hooks/useCards');

describe('CardTypes Component', () => {
    it('should display loading indicator when isLoading is true', () => {
        (useCardsModule.useCards as jest.Mock).mockReturnValue({
            cardTypes: [],
            onCardTypePress: jest.fn(),
            isLoading: true,
        });

        const { getByTestId } = render(<CardTypes />);
        expect(getByTestId('loading-indicator')).toBeTruthy();
    });

    it('should render card types when available', async () => {
        const mockCardTypes = ['Type1', 'Type2'];
        (useCardsModule.useCards as jest.Mock).mockReturnValue({
            cardTypes: mockCardTypes,
            onCardTypePress: jest.fn(),
            isLoading: false,
        });

        const { getAllByText } = render(<CardTypes />);

        await waitFor(() => {
            expect(getAllByText(/Type/i)).toHaveLength(mockCardTypes.length);
        });
    });

    it('should not render anything when there are no card types', () => {
        (useCardsModule.useCards as jest.Mock).mockReturnValue({
            cardTypes: [],
            onCardTypePress: jest.fn(),
            isLoading: false,
        });

        const { queryByText } = render(<CardTypes />);
        expect(queryByText(/Type/i)).toBeNull();
    });

    test('Cards Component renders correctly', () => {
        const {toJSON} = render(<CardTypes />);
        expect(toJSON()).toMatchSnapshot();
      });
});
