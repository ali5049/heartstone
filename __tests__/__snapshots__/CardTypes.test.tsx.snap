// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardTypes Component Cards Component renders correctly 1`] = `<RCTSafeAreaView />`;
